pkg/controller/elasticsearch/user/user_provided.go

```diff
@@ -0,0 +1,155 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package user
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/events"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/watches"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/user/filerealm"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+// UserProvidedFileRealmWatchName returns the watch registered for user-provided file realm secrets.
+func UserProvidedFileRealmWatchName(es types.NamespacedName) string {
+	return fmt.Sprintf("%s-%s-user-file-realm", es.Namespace, es.Name)
+}
+
+// UserProvidedRolesWatchName returns the watch registered for user-provided roles secrets.
+func UserProvidedRolesWatchName(es types.NamespacedName) string {
+	return fmt.Sprintf("%s-%s-user-roles", es.Namespace, es.Name)
+}
+
+// reconcileUserProvidedFileRealm returns the aggregate file realm from the referenced sources in the es spec.
+// It also ensures referenced secrets are watched for future reconciliations to be triggered on any change.
+func reconcileUserProvidedFileRealm(
+	c k8s.Client,
+	es esv1.Elasticsearch,
+	watched watches.DynamicWatches,
+	recorder record.EventRecorder,
+) (filerealm.Realm, error) {
+	esKey := k8s.ExtractNamespacedName(&es)
+	secretNames := make([]string, 0, len(es.Spec.Auth.FileRealm))
+	for _, secretRef := range es.Spec.Auth.FileRealm {
+		if secretRef.SecretName == "" {
+			continue
+		}
+		secretNames = append(secretNames, secretRef.SecretName)
+	}
+	if err := watches.WatchUserProvidedSecrets(
+		esKey,
+		watched,
+		UserProvidedFileRealmWatchName(esKey),
+		secretNames,
+	); err != nil {
+		return filerealm.Realm{}, err
+	}
+	return retrieveUserProvidedFileRealm(c, es, recorder)
+}
+
+// reconcileUserProvidedRoles returns aggregate roles from the referenced sources in the es spec.
```

Comment:
"returns aggregated roles" ?

Reply:
I asked myself the same question when writing this 😄 
From https://english.stackexchange.com/questions/241413/is-there-a-difference-in-meaning-between-aggregate-and-aggregated, it seems "aggregate" can/should be used as an adjective.

I can also see `aggregated` being used in many places, so I guess both can be used.